frame/contracts/src/storage/meter.rs

```diff
@@ -533,31 +527,70 @@ impl<T: Config> Ext<T> for ReservingExt {
 
 	fn charge(
 		origin: &T::AccountId,
-		deposit_account: &DepositAccount<T>,
+		contract: &T::AccountId,
 		amount: &DepositOf<T>,
-		terminated: bool,
+		_terminated: bool,
 	) -> Result<(), DispatchError> {
 		match amount {
 			Deposit::Charge(amount) | Deposit::Refund(amount) if amount.is_zero() => return Ok(()),
 			Deposit::Charge(amount) => {
-				T::Currency::transfer(origin, deposit_account, *amount, Preservation::Preserve)?;
-				Ok(())
+				System::<T>::inc_providers(contract);
+				// This could fail if the `origin` does not have enough liquidity. Ideally, though,
+				// this should have been checked before with `check_limit`.
+				T::Currency::transfer_and_hold(
+					&HoldReason::StorageDepositReserve.into(),
+					origin,
+					contract,
+					*amount,
+					Precision::Exact,
+					Preservation::Preserve,
+					Fortitude::Polite,
+				)?;
+
+				Pallet::<T>::deposit_event(
+					vec![T::Hashing::hash_of(&origin), T::Hashing::hash_of(&contract)],
+					Event::StorageDepositTransferredAndHeld {
+						from: origin.clone(),
+						to: contract.clone(),
+						amount: *amount,
+					},
+				);
 			},
 			Deposit::Refund(amount) => {
-				if terminated {
-					System::<T>::dec_consumers(&deposit_account);
-				}
-				T::Currency::transfer(
-					deposit_account,
+				let transferred = T::Currency::transfer_on_hold(
+					&HoldReason::StorageDepositReserve.into(),
+					contract,
 					origin,
 					*amount,
-					// We can safely make it `Expendable` because our own consumer prevents a
-					// removal.
-					Preservation::Expendable,
+					Precision::BestEffort,
+					Restriction::Free,
+					Fortitude::Polite,
 				)?;
-				Ok(())
+
+				Pallet::<T>::deposit_event(
+					vec![T::Hashing::hash_of(&contract), T::Hashing::hash_of(&origin)],
+					Event::StorageDepositTransferredAndReleased {
+						from: contract.clone(),
+						to: origin.clone(),
+						amount: transferred,
+					},
+				);
+
+				if transferred < *amount {
+					// This should never happen, if it does it means that there is a bug in the
+					// runtime logic. In the rare case this happens we try to refund as much as we
+					// can, thus the `Precision::BestEffort`.
+					log::error!(
+						target: "runtime::contracts",
+						"Failed to repatriate full storage deposit {:?} from contract {:?} to origin {:?}. Transferred {:?}.",
+						amount, contract, origin, transferred,
+					);
+				} else {
+					let _ = System::<T>::dec_providers(contract);
```

Comment:
why do we call dec_providers here?

Reply:
During the termination process we first transfer all the free balance of the contract to the beneficiary:
https://github.com/paritytech/substrate/blob/810af1c564b3b38ea91838812203751c729599e7/frame/contracts/src/exec.rs#L1277-L1282

And then we release the held balance under the `StorageDepositReserve` reason and transfer it back to the beneficiary:
https://github.com/paritytech/substrate/blob/810af1c564b3b38ea91838812203751c729599e7/frame/contracts/src/storage/meter.rs#L562-L570

If we had no provider, the former `transfer()` would not include the `ed` as it needs it to keep the account alive (even while having `Preservation::Expendable` - see [`reducible_balance`](https://github.com/paritytech/substrate/blob/810af1c564b3b38ea91838812203751c729599e7/frame/balances/src/impl_fungible.rs#L44)), and therefore we'd need to deal with this `ed` somehow later on. 
The best solution I see then is to have a provider. 

Having said that, your concerns regarding the unbalanced `inc_providers`/`dec_providers` are correct and therefore I have moved them to pretty much replace where the previous `inc_consumers`/`dec_consumers` were, to make sure we only have one of each.

Reply:
Good. This was about incrementing and decrementing for each and every storage deposit. Not that we shouldn't have a provider.